gson/src/main/java/com/google/gson/internal/bind/ObjectTypeAdapter.java

```diff
@@ -49,24 +51,41 @@ private ObjectTypeAdapter(Gson gson, ToNumberStrategy toNumberStrategy) {
     this.toNumberStrategy = toNumberStrategy;
   }
 
-  private static TypeAdapterFactory newFactory(final ToNumberStrategy toNumberStrategy) {
+  private static TypeAdapterFactory newFactory(
+      final ToNumberStrategy toNumberStrategy, final boolean skipTypeVariable) {
     return new TypeAdapterFactory() {
       @SuppressWarnings("unchecked")
       @Override
       public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
-        if (type.getRawType() == Object.class) {
+        if (type.getRawType() == Object.class
+            && (!skipTypeVariable || !isTypeVariableWithBound(type.getType()))) {
           return (TypeAdapter<T>) new ObjectTypeAdapter(gson, toNumberStrategy);
         }
         return null;
       }
+
+      private boolean isTypeVariableWithBound(Type type) {
+        if (type instanceof TypeVariable<?>) {
+          TypeVariable<?> tv = (TypeVariable<?>) type;
+          Type bound = tv.getBounds()[0];
+          return bound != Object.class && bound instanceof Class<?>;
+        } else {
+          return false;
+        }
+      }
     };
   }
 
   public static TypeAdapterFactory getFactory(ToNumberStrategy toNumberStrategy) {
+    return getFactory(toNumberStrategy, false);
+  }
+
+  public static TypeAdapterFactory getFactory(
+      ToNumberStrategy toNumberStrategy, boolean skipTypeVariable) {
     if (toNumberStrategy == ToNumberPolicy.DOUBLE) {
       return DOUBLE_FACTORY;
     } else {
```

Comment:
Is this correct? `DOUBLE_FACTORY` is created with `skipTypeVariable = true`, but here the `skipTypeVariable` argument is not checked, so it might actually be `false`.